crates/chain/src/tx_data_traits.rs

```diff
@@ -31,3 +33,79 @@ impl ForEachTxOut for Transaction {
+///
+/// This trait is typically associated with blockchain data such as transactions.
+pub trait BlockAnchor:
+    core::fmt::Debug + Clone + Eq + PartialOrd + Ord + core::hash::Hash + Send + Sync + 'static
+{
+    /// Returns the [`BlockId`] that the associated blockchain data is "anchored" in.
+    fn anchor_block(&self) -> BlockId;
+}
+
+impl<A: BlockAnchor> BlockAnchor for &'static A {
+    fn anchor_block(&self) -> BlockId {
+        <A as BlockAnchor>::anchor_block(self)
+    }
+}
+
+impl BlockAnchor for (u32, BlockHash) {
+    fn anchor_block(&self) -> BlockId {
+        (*self).into()
+    }
+}
+
+/// Represents a service that tracks the best chain history.
+/// TODO: How do we ensure the chain oracle is consistent across a single call?
+/// * We need to somehow lock the data! What if the ChainOracle is remote?
+/// * Get tip method! And check the tip still exists at the end! And every internal call
+///   does not go beyond the initial tip.
+pub trait ChainOracle {
+    /// Error type.
+    type Error: core::fmt::Debug;
+
+    /// Returns the block hash (if any) of the given `height`.
+    fn get_block_in_best_chain(&self, height: u32) -> Result<Option<BlockHash>, Self::Error>;
+
+    /// Determines whether the block of [`BlockId`] exists in the best chain.
+    fn is_block_in_best_chain(&self, block_id: BlockId) -> Result<bool, Self::Error> {
+        Ok(matches!(self.get_block_in_best_chain(block_id.height)?, Some(h) if h == block_id.hash))
+    }
+}
+
+// [TODO] We need stuff for smart pointers. Maybe? How does rust lib do this?
+// Box<dyn ChainOracle>, Arc<dyn ChainOracle> ????? I will figure it out
+impl<C: ChainOracle> ChainOracle for &C {
+    type Error = C::Error;
+
+    fn get_block_in_best_chain(&self, height: u32) -> Result<Option<BlockHash>, Self::Error> {
+        <C as ChainOracle>::get_block_in_best_chain(self, height)
+    }
+
+    fn is_block_in_best_chain(&self, block_id: BlockId) -> Result<bool, Self::Error> {
+        <C as ChainOracle>::is_block_in_best_chain(self, block_id)
+    }
+}
+
+/// Represents an index of transaction data.
+pub trait TxIndex {
+    /// The resultant "additions" when new transaction data is indexed.
+    type Additions;
+
+    /// Scan and index the given `outpoint` and `txout`.
+    fn index_txout(&mut self, outpoint: OutPoint, txout: &TxOut) -> Self::Additions;
+
+    /// Scan and index the given transaction.
+    fn index_tx(&mut self, tx: &Transaction) -> Self::Additions;
+
+    /// Apply additions to itself.
+    fn apply_additions(&mut self, additions: Self::Additions);
+
+    /// A transaction is relevant if it contains a txout with a script_pubkey that we own, or if it
+    /// spends an already-indexed outpoint that we have previously indexed.
+    fn is_tx_relevant(&self, tx: &Transaction) -> bool;
+
+    /// Returns whether the script pubkey is owned by us.
+    fn is_spk_owned(&self, spk: &Script) -> bool;
```

Comment:
Why is an index forced to provide this method?

I think you want this just for you convenience method `try_list_chain_txouts` but this is only appropriate for some indexes. You can simply implement `try_list_chain_txouts`  method where `I: AsRef<SpkTxOutIndex>`.

Reply:
I've changed the API to this:
https://github.com/evanlinjin/bdk/blob/ddd5e951f5ec77070034c7390a635d8d5bd7cb85/crates/chain/src/tx_data_traits.rs#L69-L88

I know it's probably not what you had in mind, but is it sufficient? I think this makes sense.

Reply:
It doesn't make sense because if I make an index that says that "this txout is relevant" it might not be relevant because I own that output it might be relevant because one of my transactions spends from it and I want to know the tx fee. In fact, the existing indexes should probably have an option to return true for `is_txout_relevant` in this case. The problem is that `try_list_chain_txouts` is using this trait method as if it means that someone owns the txout.

**this trait should have nothing to do with telling the indexed tx graph about which utxos you own**. Ownership semantics exists outside the `IndexedTxGraph` system. It only exists with particular indexes. Is there any reason that what I suggested doesn't work?

Reply:
Moved to a separate trait.